### ISSUE TITLE
Query Object by UDF

### ISSUE THREAD

**Issue by SirPoot-** [opendns/autotask-php#13]
First, thank you for this tool!

I have performed all the tasks; query, create, update and delete. I feel I have a good understanding on how to use your tool. I was wondering on the query, if you would query by a UDF (user defined field).

Right now, I am have a broad query then filtering those each of the results searching for a particular UDF. Then have to loop through the next board query set and so forth. I feel like there is a more efficient way, I just do not know how to do it.

Thank you in advance!

- Devin

**Comment by reynolek** [comment]
I have never actually tried this, however according to the [Autotask API documentation](https://www.autotask.net/help/content/Userguides/T_WebServicesAPIv1_5.pdf) you can query by any <entity> value:

> Permissible <entity> values: The name of any entity exposed by the Web Services API

So you could try querying by `'UserDefinedField'` -- but I have a feeling it doesnt fit the definition of entity for Autotask.

Some possible example code (again, I didn't actually run this):

```php
$query = new \ATWS\AutotaskObjects\Query('UserDefinedField');
$queryField = new \ATWS\AutotaskObjects\QueryField('NameOfYourField');
$queryField->addExpression('Equals', 10);
$query->addField($queryField);
```
And then use your autotask client to send off that query object.

I am not very confident that you will be able to do this though unfortunately, but hey, you never know!

Good Luck!

**Comment by SirPoot-** [comment]
I tried it and it did not work.

The query object is 'UserDefinedFields'.

Here is my query code
```
	$query = new \ATWS\AutotaskObjects\Query('UserDefinedFields');
	$queryField = new \ATWS\AutotaskObjects\QueryField('Opportunity Type');
	$queryField->addExpression('contains', 'BATTS');
	$query->addField($queryField);
	$results = $client->query($query);
```

Here is the response

```
object(ATWS\AutotaskObjects\QueryResponse)#9 (1) {
  ["queryResult"]=>
  object(ATWS\AutotaskObjects\ATWSResponse)#10 (5) {
    ["EntityResults"]=>
    object(ATWS\AutotaskObjects\ArrayOfEntity)#11 (1) {
      ["Entity"]=>
      NULL
    }
    ["EntityResultType"]=>
    string(17) "userdefinedfields"
    ["EntityReturnInfoResults"]=>
    object(ATWS\AutotaskObjects\ArrayOfEntityReturnInfo)#14 (3) {
      ["EntityReturnInfo"]=>
      NULL
      ["Fields"]=>
      NULL
      ["UserDefinedFields"]=>
      NULL
    }
    ["Errors"]=>
    object(ATWS\AutotaskObjects\ArrayOfATWSError)#12 (3) {
      ["ATWSError"]=>
      object(ATWS\AutotaskObjects\ATWSError)#13 (1) {
        ["Message"]=>
        string(42) "Value cannot be null.
Parameter name: type"
      }
      ["Fields"]=>
      NULL
      ["UserDefinedFields"]=>
      NULL
    }
    ["ReturnCode"]=>
    int(-1)
  }
}
  
```

UDF Info Query

```

object(ATWS\AutotaskObjects\UDFInfoResponse)#5 (1) {
  ["getUDFInfoResult"]=>
  object(ATWS\AutotaskObjects\ArrayOfField)#6 (3) {
    ["Field"]=>
    array(12) {
      [0]=>
      object(ATWS\AutotaskObjects\Field)#7 (16) {
        ["DefaultValue"]=>
        string(0) ""
        ["Description"]=>
        NULL
        ["IsPickList"]=>
        bool(true)
        ["IsQueryable"]=>
        bool(true)
        ["IsReadOnly"]=>
        bool(false)
        ["IsReference"]=>
        bool(false)
        ["IsRequired"]=>
        bool(true)
        ["Label"]=>
        string(16) "Opportunity Type"
        ["Length"]=>
        int(1024)
        ["Name"]=>
        string(16) "Opportunity Type"
        ["PicklistParentValueField"]=>
        NULL
        ["PicklistValues"]=>
        object(ATWS\AutotaskObjects\ArrayOfPickListValue)#8 (3) {
          ["PickListValue"]=>
          array(10) {
            [0]=>
            object(ATWS\AutotaskObjects\PickListValue)#9 (9) {
              ["IsActive"]=>
              bool(true)
              ["IsDefaultValue"]=>
              bool(false)
              ["IsSystem"]=>
              bool(false)
              ["Label"]=>
              string(4) "BATT"
              ["SortOrder"]=>
              int(0)
              ["Value"]=>
              string(4) "BATT"
              ["parentValue"]=>
              string(0) ""
              ["Fields"]=>
              NULL
              ["UserDefinedFields"]=>
              NULL
            }
            [1]=>
            object(ATWS\AutotaskObjects\PickListValue)#10 (9) {
              ["IsActive"]=>
              bool(true)
              ["IsDefaultValue"]=>
              bool(false)
              ["IsSystem"]=>
              bool(false)
              ["Label"]=>
              string(4) "CAPS"
              ["SortOrder"]=>
              int(1)
              ["Value"]=>
              string(4) "CAPS"
              ["parentValue"]=>
              string(0) ""
              ["Fields"]=>
              NULL
              ["UserDefinedFields"]=>
              NULL
            }
            [2]=>
            object(ATWS\AutotaskObjects\PickListValue)#11 (9) {
              ["IsActive"]=>
              bool(true)
              ["IsDefaultValue"]=>
              bool(false)
              ["IsSystem"]=>
              bool(false)
              ["Label"]=>
              string(9) "Duplicate"
              ["SortOrder"]=>
              int(2)
              ["Value"]=>
              string(9) "Duplicate"
              ["parentValue"]=>
              string(0) ""
              ["Fields"]=>
              NULL
              ["UserDefinedFields"]=>
              NULL
            }
            [3]=>
            object(ATWS\AutotaskObjects\PickListValue)#12 (9) {
              ["IsActive"]=>
              bool(true)
              ["IsDefaultValue"]=>
              bool(false)
              ["IsSystem"]=>
              bool(false)
              ["Label"]=>
              string(9) "Equipment"
              ["SortOrder"]=>
              int(3)
              ["Value"]=>
              string(9) "Equipment"
              ["parentValue"]=>
              string(0) ""
              ["Fields"]=>
              NULL
              ["UserDefinedFields"]=>
              NULL
            }
            [4]=>
            object(ATWS\AutotaskObjects\PickListValue)#13 (9) {
              ["IsActive"]=>
              bool(true)
              ["IsDefaultValue"]=>
              bool(false)
              ["IsSystem"]=>
              bool(false)
              ["Label"]=>
              string(3) "ESN"
              ["SortOrder"]=>
              int(4)
              ["Value"]=>
              string(3) "ESN"
              ["parentValue"]=>
              string(0) ""
              ["Fields"]=>
              NULL
              ["UserDefinedFields"]=>
              NULL
            }
            [5]=>
            object(ATWS\AutotaskObjects\PickListValue)#14 (9) {
              ["IsActive"]=>
              bool(true)
              ["IsDefaultValue"]=>
              bool(false)
              ["IsSystem"]=>
              bool(false)
              ["Label"]=>
              string(3) "ESR"
              ["SortOrder"]=>
              int(5)
              ["Value"]=>
              string(3) "ESR"
              ["parentValue"]=>
              string(0) ""
              ["Fields"]=>
              NULL
              ["UserDefinedFields"]=>
              NULL
            }
            [6]=>
            object(ATWS\AutotaskObjects\PickListValue)#15 (9) {
              ["IsActive"]=>
              bool(true)
              ["IsDefaultValue"]=>
              bool(false)
              ["IsSystem"]=>
              bool(false)
              ["Label"]=>
              string(3) "IRD"
              ["SortOrder"]=>
              int(6)
              ["Value"]=>
              string(3) "IRD"
              ["parentValue"]=>
              string(0) ""
              ["Fields"]=>
              NULL
              ["UserDefinedFields"]=>
              NULL
            }
            [7]=>
            object(ATWS\AutotaskObjects\PickListValue)#16 (9) {
              ["IsActive"]=>
              bool(true)
              ["IsDefaultValue"]=>
              bool(false)
              ["IsSystem"]=>
              bool(false)
              ["Label"]=>
              string(5) "Parts"
              ["SortOrder"]=>
              int(7)
              ["Value"]=>
              string(5) "Parts"
              ["parentValue"]=>
              string(0) ""
              ["Fields"]=>
              NULL
              ["UserDefinedFields"]=>
              NULL
            }
            [8]=>
            object(ATWS\AutotaskObjects\PickListValue)#17 (9) {
              ["IsActive"]=>
              bool(true)
              ["IsDefaultValue"]=>
              bool(false)
              ["IsSystem"]=>
              bool(false)
              ["Label"]=>
              string(2) "PM"
              ["SortOrder"]=>
              int(8)
              ["Value"]=>
              string(2) "PM"
              ["parentValue"]=>
              string(0) ""
              ["Fields"]=>
              NULL
              ["UserDefinedFields"]=>
              NULL
            }
            [9]=>
            object(ATWS\AutotaskObjects\PickListValue)#18 (9) {
              ["IsActive"]=>
              bool(true)
              ["IsDefaultValue"]=>
              bool(false)
              ["IsSystem"]=>
              bool(false)
              ["Label"]=>
              string(3) "T&M"
              ["SortOrder"]=>
              int(9)
              ["Value"]=>
              string(3) "T&M"
              ["parentValue"]=>
              string(0) ""
              ["Fields"]=>
              NULL
              ["UserDefinedFields"]=>
              NULL
            }
          }
          ["Fields"]=>
          NULL
          ["UserDefinedFields"]=>
          NULL
        }
        ["ReferenceEntityType"]=>
        NULL
        ["Type"]=>
        string(6) "string"
        ["Fields"]=>
        NULL
        ["UserDefinedFields"]=>
        NULL
      }
     
```

The response is mentioning the value cannot be NULL. I know UDFs are goofy where each one has a name and a value. The UDF info query is indicating my UDF 'Opportunity Type' is queryable. 

Thoughts?

**Comment by reynolek** [comment]
Unsure unfortunately -- this is something I haven't had to do in the past. Perhaps if you query for say an Account (or whatever real entity you want) and add the Opportunity Type as a query field? maybe? Thats the best guess I have.

**Comment by SirPoot-** [comment]
Thank you. I will give that a shot in the coming days. Thanks for your help!

**Comment by reynolek** [comment]
Sorry I couldn't be of more help -- If you do manage to get it to work, please comment with what you did for future people who may have the same issue!

**Comment by sulfurcode** [comment]
Not sure if this ever got resolved, but for UDF fields you have to mark it as a UDF in the query.

"To specify that a field is a UDF, you must add udf=”true” within the field tag. You can only specify one UDF field per query."

e.g 
<field udf="true">UDF NAME<expression op="Equals">1</expression></field>

If the ATWS\AutotaskObjects\QueryField() function could have another value passed called udf which defaults to false then that should allow queries on UDFs.

**Comment by reynolek** [comment]
To ensure that I understand correctly -- you want the constructor for the QueryField object to take `$isUDF = false` as an optional parameter.

And then in the xml, it should be converted to `<field udf="true">{insert field xml data}</field>`

**Comment by sulfurcode** [comment]
Yep spot on

**Comment by reynolek** [comment]
Wrote and tested this feature. Merged into master and released v0.1.5
